raft-tech/TANF-app
# Feat/2486 - Parser Performance Improvement

## Summary of Changes
- Update parsing logic to avoid multiple passes over datafile
- Update record saving to occur in batches of 10,000 instead of one at a time
Pull request closes #2486 

## How to Test

```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
cd tdrs-backend && docker-compose exec web pytest -k test_parse.py
```
- You could also verify the time decrease by running
```
cd tdrs-backend && docker-compose exec web pytest -k test_parse_ssp_section1_datafile
```
on both `develop` and this branch. On my machine, develop takes ~820 seconds to complete and this branch takes ~16 seconds to complete. Please use the attached patch to add timing code to this test. Note the test is forced to fail so that you can see the stdout of the time taken in seconds. You can apply the patch by:
```
cd tdrs-backend && git apply timer.patch
```

[timer.patch](https://github.com/raft-tech/TANF-app/files/11580133/timer.patch)


## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Identify root cause of any performance issues
- [x] Identify mitigation steps
- [x] Create a benchmarking test to compare and contrast current performance with any possible improvements

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
